src/CustomFunction.ts

```diff
@@ -1,221 +1,222 @@
-        const errorMsg = 'All CustomFunction middleware must be functions';
-        throw new CustomFunctionInitializationError(errorMsg);
-      }
-    });
+  /**
+   * Configure middleware arguments to provide to function listeners.
+   * @param context - the function execution context.
+   * @param client - a configured web client.
+   * @returns listener arguments configured for the function execution.
+   * @throws when function_execution_id is not included in arguments.
+   */
+  public static middlewareArgs(context: Context, client: WebClient): CustomFunctionMiddlewareArgs {
+    const { functionExecutionId, functionInputs } = context;
+    if (functionExecutionId === undefined) {
+      throw new CustomFunctionRuntimeError('No function_execution_id was found in the context');
+    }
+    const token = CustomFunction.selectToken(context);
+    return {
+      inputs: functionInputs ?? {},
+      complete: (params: FunctionCompleteArguments = {}): Promise<FunctionsCompleteSuccessResponse> => (
+        client.functions.completeSuccess({
+          token,
+          outputs: params.outputs ?? {},
+          function_execution_id: functionExecutionId,
+        })
+      ),
+      fail: (params: FunctionFailArguments): Promise<FunctionsCompleteErrorResponse> => (
+        client.functions.completeError({
+          token,
+          error: params.error,
+          function_execution_id: functionExecutionId,
+        })
+      ),
+    };
   }
-}
 
-/**
- * `processFunctionMiddleware()` invokes each listener middleware
- * @param args function_executed event
- */
-export async function processFunctionMiddleware(
-  args: AllCustomFunctionMiddlewareArgs,
-  middleware: CustomFunctionMiddleware,
-): Promise<void> {
-  const { context, client, logger } = args;
-  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
-  const lastCallback = callbacks.pop();
-
-  if (lastCallback !== undefined) {
-    await processMiddleware(
-      callbacks, args, context, client, logger,
-      async () => lastCallback({ ...args, context, client, logger }),
-    );
+  /**
+   * Gather a token matching the function context set during app initialization.
+   * This is either from the function execution context or a bot or user token and
+   * is the token used in `function()` handlers.
+   *
+   * The functionBotAccessToken is set to context during `App` initialization if
+   * configured during setup so that continuity between a function_executed event
+   * and subsequent interactive events (actions) is preserved.
+   *
+   * The botToken and userToken remain available in context regardless of this but
+   * the functionBotAccess token cannot be switched between functions. Middleware
+   * granularity for this decision is left TODO.
+   *
+   * @param context - the incoming payload context.
+   * @link https://github.com/slackapi/bolt-js/pull/2026#discussion_r1467123047
+   */
+  private static selectToken(context: Context): string | undefined {
```

Comment:
Great call once more! Had to think a bit about where this might make sense - exporting from `App.ts` into `CustomFunction.ts` felt strange - and ended up creating `middleware/context.ts` for similar `context` operations. Open to adjustments though :sparkles: